### ISSUE TITLE
Clustered reduced embeddings in "fit_transform" process spends a lot of time !!

### ISSUE THREAD

**Issue by apoalquaary** [MaartenGr/BERTopic#1362]
I'm making a model that contains 1.2M documents

Comparing to "Transformed documents to Embeddings" and "Reduced dimensionality", "Clustered reduced embeddings" spends a lot of time. 

Batches: 100%|██████████| 37500/37500 [06:02<00:00, 103.42it/s]
2023-06-24 20:55:40,905 - BERTopic - Transformed documents to Embeddings
2023-06-24 22:46:16,846 - BERTopic - Reduced dimensionality

as we can see "Transformed documents to Embeddings" and "Reduced dimensionality" spent almost (6m, 2h). now it passed more than 15h and "Clustered reduced embeddings" didn't finish yet. Is there nested loops or what do you think the reason is?

**Comment by MaartenGr** [comment]
It is difficult to say without seeing the full code. Could you share it? Also, which version of BERTopic are you using? 

Generally, it is the result of the CPU version of HDBSCAN to take quite a bit of time to cluster the documents especially if you set `calculate_probabilities=True`. If you want to speed up clustering, using cuML's [HDBSCAN](https://maartengr.github.io/BERTopic/getting_started/clustering/clustering.html#cuml-hdbscan) instead to accelerate the model using a GPU.

**Comment by apoalquaary** [comment]
sentence_model = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2")
umap_model = UMAP(n_neighbors=15, n_components=200,
                  min_dist=0.0, metric='cosine', random_state=100)

ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)
representation_model = KeyBERTInspired()

topic_model = BERTopic(language="multilingual", umap_model = umap_model, verbose=True, n_gram_range = (1,1), representation_model=representation_model, ctfidf_model=ctfidf_model, top_n_words=10)
topics, probs = topic_model.fit_transform(docs) # _transform 

this is the code .. should I write specific thing to run HDBSCAN on GPU

**Comment by apoalquaary** [comment]
"I greatly appreciate your assistance and the valuable response. I have a question regarding the selection of suitable hyperparameters for a large-scale dataset comprising millions of microblogging data in diverse languages."


sentence_model = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2")
umap_model = UMAP(n_neighbors=15, n_components=15,
                  min_dist=0.0, metric='cosine', random_state=100)
hdbscan_model = HDBSCAN(min_cluster_size=10, min_samples = 10, metric='euclidean', prediction_data=True)

ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)
representation_model = KeyBERTInspired()

Do you have any comment about this.

**Comment by MaartenGr** [comment]
I would advise using the GPU-accelerated version of HDBSCAN for the clustering as that will speed up things significantly.

**Comment by MaartenGr** [comment]
Closing this due to inactivity. Let me know if you want to re-open the issue!
